[auth]: add authentication

legomushroom opened the pull request:
0. Adds the required Bearer auth strategy service API to the endpoints.
1. Adds the AAD JWT verification logic.
1. Adds the Cascade JWT verification logic.
2. Moves the extension to use the `@vs/vscode-account` module for authentication to share the access tokens with Live Share.

Still waits on:

0. Live Share change and release to reuse the auth strategies used by Live Share.

legomushroom commented on service/.vscode/launch.json:
```diff
@@ -0,0 +1,29 @@
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
```
Remove.

legomushroom commented on service/.vscode/tasks.json:
```diff
@@ -0,0 +1,10 @@
+{
+    "version":"2.0.0",
```
Remove.

legomushroom commented on service/mix.exs:
```diff
@@ -27,7 +27,9 @@ defmodule LiveShareCommunities.MixProject do
       {:poison, "~> 4.0.1"},
       {:redix, ">= 0.0.0"},
-
+      {:joken, "~> 2.0-rc0"},
```
Remove `joken`.

arjun27 replied:
minor: removing dependency from mix.exs does not remove it from the lock file. that needs an additional `mix deps.unlock --unused`

legomushroom commented on src/api.ts:
```diff
@@ -1,80 +1,128 @@
 
-const BASE_URL = "http://vslscommunitieswebapp.azurewebsites.net/v0";
+// const BASE_URL = "http://vslscommunitieswebapp.azurewebsites.net/v0";
+const BASE_URL = "http://localhost:4000/v0";
```
Return the old one.

arjun27 commented:
looks awesome! would we want auth on the websocket?

also, couple of very minor things
- the dump.rdb file got committed (backed up from running Redis-server)
- can we remove package-lock.json (using yarn)?

arjun27 commented on .npmrc:
```diff
@@ -0,0 +1,2 @@
+@vs:registry=https://devdiv.pkgs.visualstudio.com/_packaging/VS/npm/registry/
```
would this private feed need a PAT? how will this work with deploy.sh?

legomushroom replied:
Yep, this means you will have to connect to the feed @VS feed at azure devops. If it deployed from the Azure devops it should be authenticated there. How do we deploy right now?

arjun27 replied:
at this stage I'm running `sh deploy.sh` locally. I'm also trying to get us access to the new GitHub Actions so we can set up that for CI/CD

legomushroom replied:
If you run deploy locally and build the bits on your machine then it will work (if you will connect to the feed).

legomushroom commented:
@arjun27 @lostintangent added the Cascade token validation, now the user will be able to use any (ms/gh) account thru live share.
Also addressed some feedback please take a look.


Re the WS auth, what are we using the WS for at the moment?

arjun27 commented:
> Re the WS auth, what are we using the WS for at the moment?

We're using the websocket transport that comes with [cowboy](https://github.com/ninenines/cowboy) (webserver for elixir) – cowboy is also listening for http. We could de-prioritize gating on websocket, at least till we implement private communities (#45)

Code for websocket, on the service, in case you want to review: https://github.com/vsls-contrib/communities/blob/master/service/lib/communities/websocket.ex

legomushroom commented:
@arjun27 ya, but what scenarios we are using the WS for?

arjun27 commented:
> what scenarios we are using the WS for?

The websocket delivers real-time updates for: changes to the communities the user is a member of (member added/removed, founder changed), and new requests and messages in these communities

legomushroom commented:
@arjun27 I see, thanks. Sounds like we do want to restrict those also. Maybe not in this PR though.

lostintangent commented on service/lib/communities/auth.ex:
```diff
@@ -0,0 +1,283 @@
+
+  def init(opts), do: opts
+
+  defmemo getKeys?(arg) do
```
Should we make all of these function names use snake case? (E.g. “get_keys?”). That seems to be the convention in Elixir.

lostintangent commented on src/config.ts:
```diff
@@ -18,6 +19,8 @@ export const config = {
       );
     }
+
+    await vscode.commands.executeCommand('liveshare.enable.vscode-account.auth');
```
This file has been updated to import the `commands` API already, so you should pull the latest changes from `MASTER` and should be able to just call `commands.executeCommand('liveshare.enable.vscode-account.auth')`.

legomushroom replied:
Done!

lostintangent commented on .vscode/launch.json:
```diff
@@ -3,34 +3,34 @@
-		}
-	]
-}
+    "version": "0.2.0",
```
If this is just formatting, can you revert your changes? We made some changes to this file in `MASTER`.

legomushroom replied:
I took all the incoming, didnt make any changes to the file really.
